### ISSUE TITLE
EOFError not being passed to error_handler in SFTPClient.get()

### ISSUE THREAD

**Issue by geraldlnj** [ronf/asyncssh#313]
Hi, I'm trying to download files using `SFTPClient.get()`, and I pass in an error_handler and progress_handler. I'm testing it out with a bunch of intended errors.

The 2 files I have with errors are `1.txt` (insufficient permissions) and `2.txt` (does not exist).

1.txt gives me a "Permission denied" error message as expected, but the error arising from 2.txt does not seem to be passed to the error_handler. Instead, in `SFTPHandler.recv_packets()`, it encounters the EOFError with message `0 bytes read on a total of 4 expected bytes`, which calls the `self._cleanup()` method, terminating the connection.

If I swap the order of errors (missing file, followed by insufficient permissions), the connection is closed at the EOFError and does not proceed to the permission error. 

Shouldn't this EOFError be passed to error_handler instead?

This is how I'm calling SFTP.get():
```python
try:
    async with asyncssh.connect(self.remote_host, port=self.port, options=self.ssh_options) as conn:
        async with conn.start_sftp_client() as sftp:
            for dest, src in source_dict.items():
                Path(dest).mkdir(parents=True, exist_ok=True)
                await sftp.get(src,
                    localpath=dest,
                    progress_handler=progress_handler,
                    error_handler=error_handler)
except Exception as e:
    self.log.error(e)
```

**Comment by ronf** [comment]
If you are getting an EOFError in recv_packets(), that indicates that the underlying SSHChannel (or possibly the entire SSH connection) is being closed on you. Such errors are out of scope for the error_handler, which is only meant to be called when getting I/O errors either on local files or I/O errors reported back in SFTP responses. An error in the SSH connection basically means the SFTP client is no longer usable, and so that's raised as a regular exception.

It sounds to me like whatever you are using as an SFTP server is choosing to kill the SFTP session when the file doesn't exist. That doesn't sound right to me, but you'd need to check the logs on the SFTP server to see what's actually triggering that.

Are you using AsyncSSH as the SFTP server, or something else? Can you enable debug logging on it?

**Comment by geraldlnj** [comment]
I don't think it's an error with the SFTP server. I tested a small script to pull 2 files from test.rebex.net, `abc.txt` which does not exist, and `pub/example/readme.txt` which does.

```python
import asyncio, asyncssh, sys

def progress_handler(source_path, destination_path, uploaded_bytes, total_bytes):
    print('success!')

def error_handler(error):
    print('error!')

async def run_client():
    async with asyncssh.connect('test.rebex.net', username='demo', password="password") as conn:
        async with conn.start_sftp_client() as sftp:
            await sftp.get(
                ['abc.txt', 'pub/example/readme.txt'],
                localpath='.',
                progress_handler=progress_handler,
                error_handler=error_handler)

try:
    asyncio.get_event_loop().run_until_complete(run_client())
except (OSError, asyncssh.Error) as exc:
    sys.exit(str(exc))
```

I should expect to see 
```
error!
success!
```

but I just get:
```
File not found.
```

**Comment by ronf** [comment]
Thanks very much for the test program. Looking at the debug output, I see that the EOF is being initiated by this client code. I need to examine the code a bit more, but I think you may be right that it is not treating the 'File not found' error as something to pass to the error handler, and as a result it is aborting the entire get() operation with that error without trying the remaining files.

I'll get back to you once I have more information, but if that's the case I agree it should be changed.

**Comment by ronf** [comment]
Good news - I think I've found the issue. It happens when you pass in a source path that doesn't exist or otherwise triggers an error in the initial stat()/lstat() call, when the SFTP client is trying to figure out whether the path is a plain file, directory, or symlink to get()/put()/copy(). That initial stat call was outside of the "try" block for using the error handler. The fix looks pretty simple:

```patch
--- a/asyncssh/sftp.py
+++ b/asyncssh/sftp.py
@@ -2054,12 +2054,12 @@ class SFTPClient:
                     progress_handler, error_handler):
         """Copy a file, directory, or symbolic link"""
 
-        if follow_symlinks:
-            srcattrs = await srcfs.stat(srcpath)
-        else:
-            srcattrs = await srcfs.lstat(srcpath)
-
         try:
+            if follow_symlinks:
+                srcattrs = await srcfs.stat(srcpath)
+            else:
+                srcattrs = await srcfs.lstat(srcpath)
+
             if stat.S_ISDIR(srcattrs.permissions):
                 if not recurse:
                     raise SFTPFailure('%s is a directory' %
```

I tried this change on the test program above, and it now works for me. Can you confirm if this fixes the problem for you? If so, I'll go ahead and get this into the "develop" branch right away.

**Comment by geraldlnj** [comment]
Yup, it works great! Should have spotted that, I was even logging the value of `srcattrs`! Thanks!

**Comment by ronf** [comment]
Thanks for confirming the fix! This change is now available as commit 5fb2854 in the AsyncSSH "develop" branch and will be included in the next release.

**Comment by ronf** [comment]
This fix is now available in AsyncSSH 2.4.2.
